kdc/windc.c

```diff
@@ -202,3 +206,14 @@ _kdc_check_access(krb5_context context,
 			       req->msg_type == krb_as_req);
     return ret;
 }
+
+/*
+ * Instance value should capture all libraries that export data structures
+ * that are used by the plugin so, in this case, libkdc, libhdb and libkrb5.
+ */
+uintptr_t
+kdc_get_instance(void)
```

Comment:
I would prefer this take a `const char *libname` and return the corresponding library's "instance".

Reply:
How would you envisage this working with a common plugin load/initialize API between subsystems, i.e. where the plugin can only return a single instance to the consuming API?

Also, where should that API live? If it's in libkrb5, then it can't link against libkdc or libkadm5 (for example), in order to return those instances.

Reply:
Goals were:

* a single plugin load/initialize API for all (well, most) plugins – i.e. libkrb5, libkdc, libhdb, libkadm5
* instance validation done by the consuming library so plugins cannot bypass it (i.e. plugin returns instance value at load time)
* no change in existing library dependencies
* instance value (attempts to) capture instances of dependent libraries, where dependent is defined as being all libraries that expose structures to the plugin (as opposed to the actual library dependency chain). Hence XOR.

Reply:
I suppose we could have a _krb5_register_instance() API that other APIs could use. But I kind of like the idea that the plugin only has to lookup one cookie to return to the loader.

Reply:
OK, I have some ideas...

Reply:
Either the caller, or the plugin, should give the other a function _pointer_ of a get-instance function (not some value returned by such a function).  Then the one receiving it can call it and check that the instances match.

Reply:
OK, the plugin load function now looks like this –

```c
typedef krb5_error_code
(KRB5_CALLCONV *krb5_plugin_load_t)(krb5_context context,
                                    krb5_get_instance_func *func,
                                    size_t *n_deps,
                                    const char *const **deps,
                                    size_t *n_plugins,
                                    const common_plugin_ftable *const **plugins);
```

It's a bit unwieldy, but it works. Other options would be to just have the caller (consumer) provide a get_instance function and have the plugin do the validation itself; I mean, it can avoid any validation by omitting an instance list (well, almost, we will always check krb5).

Reply:
But I think providing a list of deps is less error prone than copying boilerplate code to check the instances.

Reply:
Fixed

```c
KRB5_LIB_FUNCTION uintptr_t KRB5_LIB_CALL
krb5_get_instance(krb5_context context, const char *module);
```